pymemcache/client/hash.py

```diff
@@ -0,0 +1,282 @@
+                       class. ``max_pool_size`` and ``lock_generator`` can
+                       be used with this. default: False
+
+          retry_attempts: Amount of times a client should be tried before it
+                          is marked dead and removed from the pool.
+          retry_timeout: Time in seconds that should pass between retry
+                         attempts.
+          dead_timeout: Time in seconds before attempting to add a node back in
+                        the pool.
+
+        Further arguments are interpreted as for :py:class:`.Client`
+        constructor.
+        """
+        self.clients = {}
+        self.retry_attempts = retry_attempts
+        self.retry_timeout = retry_timeout
+        self.dead_timeout = dead_timeout
+        self.use_pooling = use_pooling
+        self.key_prefix = key_prefix
+        self._failed_clients = {}
+        self._dead_clients = {}
+        self._last_dead_check_time = time.time()
+
+        if hasher is None:
+            self.hasher = RendezvousHash()
+
+        self.default_kwargs = {
+            'connect_timeout': connect_timeout,
+            'timeout': timeout,
+            'no_delay': no_delay,
+            'socket_module': socket_module,
+            'key_prefix': key_prefix,
+        }
+
+        if serializer is not None:
+            self.default.kwargs.update({
+                'serializer': serializer.serialize,
+                'deserializer': serializer.deserialize,
+            })
+
+        if use_pooling is True:
+            self.default_kwargs.update({
+                'max_pool_size': max_pool_size,
+                'lock_generator': lock_generator
+            })
+
+        for server, port in servers:
+            self.add_server(server, port)
+
+    def add_server(self, server, port):
+        key = '%s:%s' % (server, port)
+
+        if self.use_pooling:
+            client = PooledClient(
+                (server, port),
+                **self.default_kwargs
+            )
+        else:
+            client = Client((server, port))
+
+        self.clients[key] = client
+        self.hasher.add_node(key)
+
+    def remove_server(self, server, port):
+        dead_time = time.time()
+        self._failed_clients.pop((server, port))
+        self._dead_clients[(server, port)] = dead_time
+        key = '%s:%s' % (server, port)
+        self.hasher.remove_node(key)
+
+    def _get_client(self, key):
+        _check_key(key, self.key_prefix)
+        if len(self._dead_clients) > 0:
+            current_time = time.time()
+            ldc = self._last_dead_check_time
+            # we have dead clients and we have reached the
+            # timeout retry
+            if current_time - ldc > self.dead_timeout:
+                for server, dead_time in self._dead_clients.items():
+                    if current_time - dead_time > self.dead_timeout:
+                        print('bringing server back in rotation')
```

Comment:
Make this a log statement instead of a print to avoid contaminating stdout (or at least print to sys.err).